keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,757 @@
+scripts are simple Bash or Python which just simply glue a few `kubectl` commands. It is too heavy and "not worth" 
+for user to refactor just "a few commands" into "a batch of complex code". So we are seeking an optimization for `kubectl`. With this optimization, all users of kubectl can get free performance improvement.
+
+
+
+### Goals
+
+* Introduce WatchList to `kubectl` without breaking any user interface.
+* Make `kubectl` hit the cache, thus reducing impact on `kube-apiserver`.
+
+### Non-Goals
+
+* Change kubectl functionality.
+
+## Proposal
+
+`kubectl` commands like `get`, `describe` now should be able to use `WatchList` to list resources, rather than
+directly list resources page by page.
+
+At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
+to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.
+
+<<[UNRESOLVED Will `WatchList` replace `List` completely]>>
+`WatchList` was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
+over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
+rely on it without other fallback mechanisms.
+
+In my view, due to the performance improvements, we should use `WatchList` by default instead of `List` in the
+long term. For the evolution strategy, we can follow the approach of `client-go` and address it in a later release.
+
+<<[/UNRESOLVED]>>
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I shouldn't need to worry about the underlying details. I should simply update the version of `kubectl`
+to gain the performance improvements for free.
+
+#### Story 2
+
+As a cluster administrator, after upgrading the cluster, I should observe a significant reduction in the requests
+penetrating into etcd on the monitoring dashboard. Ideally, the majority of the requests should be cached by the 
+apiserver, rather than penetrating into etcd.
+
+### Risks and Mitigations
+
+For users, there are no visible changes to the user interface. However, since different APIs are called, 
+some users with restricted permissions may encounter authorization issues. Specifically, users might need
+to be granted additional `watch` permissions for the corresponding resources to ensure this mechanism 
+works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
+additional permissions, the old behavior should still function properly. In the worst case, some extra API 
+calls might cause a slight delay.
+
+Regarding security, users who already have `list` permissions inherently have access to all objects of this type. 
+Therefore, I don't believe this change will introduce any additional security issues.
+
+## Design Details
+
+Since the WatchList feature was merged in version 1.27, it allows us to include the sendInitialEvents parameter
```

Comment:
```suggestion
Since support for watch lists was merged in version 1.27, the API server allows clients to include the `sendInitialEvents` query parameter
```